doc/source/advancedexamples.rst

```diff
@@ -950,3 +950,89 @@ Note that the ``minimize`` method optimizes both the free parameters ``p`` of
 for the total evolution time should be the last value of the given
 ``initial_guess`` array.
+
+
+.. _symbolicham-example:
+
+How to define custom Hamiltonians using symbols?
+------------------------------------------------
+
+In order to use the VQE, QAOA and time evolution models in Qibo the use has to
+define Hamiltonians based on :class:`qibo.base.hamiltonians.Hamiltonian`, or
+:class:`qibo.base.hamiltonians.TrotterHamiltonian` when the Trotter
+decomposition is to be used. Qibo provides pre-coded Hamiltonians for some
+common physics models, such as the transverse field Ising model (TFIM) and the
+Heisenberg model. In order to explore other problems the user needs to define
+the Hamiltonian objects from scratch.
+
+A standard way to define Hamiltonians is through their full matrix
+representation. For example the following code generates the TFIM Hamiltonian
+with periodic boundary conditions for four qubits by constructing the
+corresponding 16x16 matrix:
+
+.. code-block::  python
+
+    import numpy as np
+    from qibo import hamiltonians, matrices
+
+    # ZZ terms
+    matrix = np.kron(np.kron(matrices.Z, matrices.Z), np.kron(matrices.I, matrices.I))
+    matrix += np.kron(np.kron(matrices.I, matrices.Z), np.kron(matrices.Z, matrices.I))
+    matrix += np.kron(np.kron(matrices.I, matrices.I), np.kron(matrices.Z, matrices.Z))
+    matrix += np.kron(np.kron(matrices.Z, matrices.I), np.kron(matrices.I, matrices.Ζ))
+    # X terms
+    matrix += np.kron(np.kron(matrices.X, matrices.I), np.kron(matrices.I, matrices.I))
+    matrix += np.kron(np.kron(matrices.I, matrices.X), np.kron(matrices.I, matrices.I))
+    matrix += np.kron(np.kron(matrices.I, matrices.I), np.kron(matrices.X, matrices.Ι))
+    matrix += np.kron(np.kron(matrices.I, matrices.I), np.kron(matrices.I, matrices.X))
+    # Create Hamiltonian object
+    ham = hamiltonians.Hamiltonian(4, matrix)
+
+
+Although it is possible to generalize the above construction to arbitrary number
+of qubits this procedure may be more complex for other Hamiltonians. Moreover
+constructing the full matrix does not scale well with increasing the number of
+qubits. This makes the use of :class:`qibo.base.hamiltonians.TrotterHamiltonian`
+preferrable as the qubit number increases, however these Hamiltonians are harder
+to construct.
+
+To simplify the construction of arbitrary Hamiltonians, Qibo provides the
+:meth:`qibo.base.hamiltonians.Hamiltonian.from_symbolic` and
+:meth:`qibo.base.hamiltonians.TrotterHamiltonian.from_symbolic` methods which
+allow the user to construct Hamiltonian objects just by writing their symbolic
+form using ``sympy`` symbols. For example, the TFIM on an arbitrary number of
+qubits ``nqubits`` could be constructed as:
+
+.. code-block::  python
+
+    import sympy
+    import numpy as np
+    from qibo import hamiltonians, matrices
+
+    # Define symbols for X and Z operators
+    x_symbols = sympy.symbols(" ".join((f"X{i}" for i in range(nqubits))))
```

Comment:
The nqubits definition is missing.

Reply:
Following our discussion yesterday, I changed this example to be specific for four qubits. This avoids the strange `join` manipulations of strings and should make it more accessible to beginners. I also changed notations slightly and now the `sympy` construction is pretty close to the "paper" form of the Hamiltonian.